cgi-bin/LJ/Talk.pm

```diff
@@ -2549,6 +2545,292 @@ sub enter_imported_comment {
+
+    my $tp_d = '/talkpost_do.tt'; # for ml strings
+
+    # Commenter can be undef for anon, but yes, we absolutely need an entry.
+    croak("Need LJ::Entry object to reply to") unless $entry->isa('LJ::Entry');
+
+    # For most errors, report and keep going; we'll return undef at the end, and
+    # the user can address them all at once. ~But if it's all gone wrong &
+    # there's nothing left to learn: go ahead and return~ (guitar)
+    my $err = sub {
+        my $error = shift;
+        push @$errret, $error;
+        return undef;
+    };
+    my $mlerr = sub {
+        return $err->( LJ::Lang::ml(@_) );
+    };
+
+    my $journalu = $entry->journal;
+    my $commenter_is_user = LJ::isu($commenter);
+
+    # First: accept the things u cannot change. Existential errors a commenter
+    # can't do anything about.
+
+    # Can the user even view this post?
+    unless ( $entry->visible_to($commenter) ) {
+        $mlerr->("$tp_d.error.mustlogin") unless $commenter_is_user;
+        $mlerr->("$tp_d.error.noauth");
+        return undef; # Shouldn't tell you anything else about this entry, then.
+    }
+    # No replying to readonly/locked/expunged journals
+    return $mlerr->("$tp_d.error.noreply_readonly_journal") if $journalu->is_readonly;
+    return $mlerr->('talk.error.purged') if $journalu->is_expunged;
+    return $err->("Account is locked, unable to post or edit a comment.") if $journalu->is_locked;
+    # can ANYONE comment?
+    return $mlerr->("$tp_d.error.nocomments") if $entry->comments_disabled;
+    # no replying to suspended entries, even by entry poster
+    return $mlerr->("$tp_d.error.noreply_suspended") if $entry->is_suspended;
+    # check max comments (unless editing existing comment)
+    if ( ! $content->{editid} && over_maxcomments( $journalu, $entry->jitemid ) ) {
+        return $mlerr->( "$tp_d.error.maxcomments" );
+    }
+    # If replying to a comment, it's gotta exist. (Hold onto it, we'll want it later.)
+    my $parenttalkid = ($content->{parenttalkid} || $content->{replyto} || 0) + 0;
+    my $parpost;
+    if ($parenttalkid) {
+        my $dbcr = LJ::get_cluster_def_reader($journalu);
+        return $mlerr->('error.nodb') unless $dbcr; # tbh we got bigger problems at this point.
+        $parpost = LJ::Talk::get_talk2_row( $dbcr, $journalu->{userid}, $parenttalkid );
+        unless ($parpost) {
+            return $mlerr->("$tp_d.error.noparent");
+        }
+    }
+    # no replying to frozen comments
+    my $parent_state = $parpost->{state} // '';
+    return $mlerr->( "$tp_d.error.noreply_frozen" ) if $parent_state eq 'F';
+
+    # Next: Permissions checks! Easily solved, just become someone else.
+
+    # (For switching between variant error messages.)
+    my $iscomm = $journalu->is_community ? '.comm' : '';
+
+    if ( $commenter_is_user ) {
+
+        # test accounts can only comment on other test accounts.
+        if (   ( grep { $commenter->user eq $_ } @LJ::TESTACCTS )
```

Comment:
Wow what is this functionality! Pretty sure we don't use TESTACCTS.

Reply:
I've never seen it before but it's referenced in a few other places too.

Reply:
yeah..... for returning a message that doesn't seem to exist from LJ::Protocol's login method, and for complaining at you if you try to change the test account's password in DW::Controller::Settings::changepassword_handler. Seems, uh,

Reply:
Totally OK!

Reply:
I feel like it's some kind of relic of LJ:IS_DEV_SERVER somehow?